### PR TITLE
feat($animateRunner): prototype code for $animateRunner

### PR DESCRIPTION
**Pull request by matsko**
DO NOT MERGE. JUST A PR FOR REVIEW PURPOSES.

### CODE REVIEW

File: src/ngAnimate/animateRunner.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+var $$rAFMutexFactory = ['$$rAF', function($$rAF) {
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: Not sure about calling this a Mutex...

File: src/ngAnimate/animateRunner.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      passed ? fn() : $$rAF(fn);
+    };
+  }
+}];
+
+var $AnimateRunnerFactory = ['$q', '$$rAFMutex', function($q, $$rAFMutex) {
+  var DONE_PENDING_STATE = 1;
+  var DONE_COMPLETE_STATE = 2;
+
+  AnimateRunner.chain = function(chain, callback) {
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: This is a function to run a sequence of functions in series right?

File: src/ngAnimate/animateRunner.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+      chain[index](function(response) {
+        status = status && response;
+        index++;
+        next();
+      });
+    }
+  };
+
+  AnimateRunner.all = function(runners, callback) {
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: Looks good

File: src/ngAnimate/animateRunner.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    this._doneCallbacks = [];
+    this._atleastOneRAFPassedSinceStart = $$rAFMutex();
+    this._state = 0;
+  };
+
+  AnimateRunner.prototype = {
+    setHost: function(host) {
+      this.host = host || {};
+    },
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: So this is where the host gets added at some point after the runner was created, right?
Looks good

File: src/ngAnimate/animateRunner.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        callback(status);
+      }
+    }
+  };
+
+  function AnimateRunner(host) {
+    this.setHost(host);
+
+    this._doneCallbacks = [];
+    this._atleastOneRAFPassedSinceStart = $$rAFMutex();
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: I would just call this something like `batchCall` or `runInAnimationFrame`. The fact that it uses rAF does not need to be exposed.

File: src/ngAnimate/animateRunner.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    },
+
+    cancel: function() {
+      this.host.cancel && this.host.cancel();
+      this._resolve(false);
+    },
+
+    complete: function(response) {
+      var self = this;
+      if (self._state === 0) {
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: Use a constant for `0` here, perhaps `INITIAL_STATE`

File: src/ngAnimate/animateRunner.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      }
+    }
+  };
+
+  function AnimateRunner(host) {
+    this.setHost(host);
+
+    this._doneCallbacks = [];
+    this._atleastOneRAFPassedSinceStart = $$rAFMutex();
+    this._state = 0;
```
Comment (petebacondarwin) [inline src/ngAnimate/animateRunner.js]: Use a constant here, maybe `INITIAL_STATE`

**Comment by petebacondarwin** [comment]
I haven't had time to look at the tests but the main code looks very nice.

File: src/ngAnimate/animateRunner.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        self._state = DONE_PENDING_STATE;
+        self._atleastOneRAFPassedSinceStart(function() {
+          self._resolve(response);
+        });
+      }
+    },
+
+    _resolve: function(response) {
+      if (this._state !== DONE_COMPLETE_STATE) {
+        angular.forEach(this._doneCallbacks, function(fn) {
```
Comment (gkalpak) [inline src/ngAnimate/animateRunner.js]: `angular.` can be ommitted (for consistency).

File: src/ngAnimate/animateRunner.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      }
+      return this.promise;
+    },
+
+    then: function(resolveHandler, rejectHandler) {
+      return this.getPromise().then(resolveHandler, rejectHandler);
+    },
+
+    pause: function() {
+      this.host.pause && this.host.pause();
```
Comment (gkalpak) [inline src/ngAnimate/animateRunner.js]: I think this type statements are discouraged (if not "frown upon") inside the Angular codebase.
Just saying... :)

File: test/ngAnimate/animateRunnerSpec.js [hunk 10]
```diff
@@ -0,0 +1,251 @@
+'use strict';
+
+describe("$$rAFMutex", function() {
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: No, no, no, no !
No !
No, no annnd no !

Consistent use of `'`, please :stuck_out_tongue:

File: test/ngAnimate/animateRunnerSpec.js [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    var capturedValue;
+    runner.done(function(val) {
+      capturedValue = val;
+    });
+    runner.complete('special value');
+    $$rAF.flush();
+    expect(capturedValue).toBe('special value');
+  }));
+
+  they("should immediately resolve each combined runner in depth-first order when $prop is called",
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: I don't think `depth-first` is an accurate description. You probably meant something like `bottom-up` or `fifo` or something.

File: test/ngAnimate/animateRunnerSpec.js [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+      runner2.then(
+        function() { status2 = 'pass'; },
+        function() { status2 = 'fail'; });
+
+      runner1[method]();
+
+      var expectedStatus = method === 'end' ? 'pass' : 'fail';
+
+      expect(status1).toBeFalsy();
+      expect(status2).toBeFalsy();
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: Aren't they expected to be undefined ? If so, using `toBeUndefined()` would be easier to follow.
(`toBeFalsy()` had me wondering if I had missed something.)

File: test/ngAnimate/animateRunnerSpec.js [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+  they("should resolve/reject using a newly created promise when .then() is used upon $prop",
+    ['end', 'cancel'], function(method) {
+
+    inject(function($animateRunner, $rootScope) {
+      var runner1 = new $animateRunner();
+      var runner2 = new $animateRunner();
+      runner1.setHost(runner2);
+
+      var status1;
+      runner2.then(
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: Was this supposed to be `runner1` ?

File: test/ngAnimate/animateRunnerSpec.js [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+      expect(status1).toBe(expectedStatus);
+      expect(status2).toBe(expectedStatus);
+    });
+  });
+
+  it("should expose/create the contained promise when getPromise() is called",
+    inject(function($animateRunner, $rootScope) {
+
+    var runner = new $animateRunner();
+    expect(isPromiseLike(runner.getPromise()));
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: I don't think you are testing anything (`.toBeTruthy()` missing ?).
<sub>(At least Jasmine 2.x won't verify anything just by calling `expect()`. Too lazy to try with 1.x :D)</sub>

File: test/ngAnimate/animateRunnerSpec.js [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+      runner3.complete(true);
+
+      expect(status).toBeUndefined();
+
+      $$rAF.flush();
+
+      expect(status).toBe(true);
+    }));
+
+    they("should immediately resolve if and when all runners have resolved been $prop",
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: _have been $prop_ ?

File: test/ngAnimate/animateRunnerSpec.js [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+
+        runner1[method]();
+        runner2[method]();
+        runner3[method]();
+
+        expect(status).toBe(expectedStatus);
+      });
+    });
+
+    it("should return a status of `false` if more than one runner was cancelled",
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: You probably mean _if one or more_.
(_"more than one"_ means _at least 2_.)

File: test/ngAnimate/animateRunnerSpec.js [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+      items.push(function(fn) { log.push(3); runner3.done(fn); });
+
+      var status;
+      $animateRunner.chain(items, function(response) {
+        status = response;
+      });
+
+      runner1.complete(true);
+      runner2.complete(true);
+      runner3.complete(true);
```
Comment (gkalpak) [inline test/ngAnimate/animateRunnerSpec.js]: This does not make it clear (nor tests) what happens if the runners are complete in a different order than the order they appear in `items`.

Would it be better/more explicit to `complete()` them in a different order and verify that the log order still matches their order in `items` ?

**Comment by gkalpak** [comment]
I really like where the whole "animations in Angular" story is going !
I am pretty sure @matsko will put this `$animateRunner` in good use for us :)

I have left some comments inline. (Nothing major, apart from the `'`/`"` inconsistency :P)

--
Some final thoughts/questions:

1. Angular is supposed to:
    * _**prefer simple objects**, instead of complex inheritance hierarchies_
    * _use prototypal inheritance only when absolutely necessary_
    * _**love functions and closures**_

  I know there isn't any _complex inheritance hierarchy_ here, but still wondering if there would be any benefit in using closures instead (probably not).

2. I understand this is not merge-ready yet, but (before merging) it would be nice to have more docs/comments (not necessarily public) on what each method does and what purpose it serves. It would be much easier to maintain/enhance. 
<sub>`ngAnimate` is pretty much a "one-matsko-show" right now, but this might (have to) change soon, considering the increasing complexity of animations and the rapidly expading "surface" they need to cover (Material Design/ngMaterial, Angular2, need for complex animation choreographies etc).</sub>

File: src/ngAnimate/animateRunner.js (hunk 9 above)
Comment (matsko) [inline src/ngAnimate/animateRunner.js]: I would prefer (fn || noop) but it doesn't work for some reason.

**Comment by matsko** [comment]
@gkalpak thanks a bunch for the great work with the review. I've made the changes, with the exception of the last one. I think I will redo that test.

**Comment by matsko** [comment]
Closed in favour of #11336.

**Pull request closed without merging**